tests/core_test.py

```diff
@@ -303,5 +303,41 @@ def test_var_tree_flatten(self):
     syms = {c: d, a: b}
     assert 'bd' == ''.join(map(str, tree_leaves(syms)))
 
+  def test_check_jaxpr_correct(self):
+    jaxpr = make_jaxpr(lambda x: np.sin(x) + np.cos(x))(1.).jaxpr
+    core.check_jaxpr(jaxpr)
+
+  def test_check_jaxpr_eqn_mismatch(self):
+    def f(x):
+      return np.sin(x) + np.cos(x)
+
+    def new_jaxpr():
+      return make_jaxpr(f)(1.).jaxpr
+
+    # jaxpr is:
+    #
+    # { lambda  ; a.
+    #   let b = sin a
+    #       c = cos a
+    #       d = add b c
+    #   in (d,) }
+    #
+    # NB: eqns[0].outvars[0] and eqns[2].invars[0] are both 'b'
+
+    jaxpr = new_jaxpr()
+    jaxpr.eqns[0].outvars[0].aval = make_shaped_array(2)   # int, not float!
+    jtu.check_raises_regexp(
+        lambda: core.check_jaxpr(jaxpr),
+        TypeError, ("Jaxpr equation LHS .* is ShapedArray(.*), "
+                    "RHS is inferred as ShapedArray(.*), in '.* = sin .*'"))
+
+    jaxpr = new_jaxpr()
+    jaxpr.eqns[0].outvars[0].aval = make_shaped_array(np.ones((2, 3)))
+    jtu.check_raises_regexp(
+        lambda: core.check_jaxpr(jaxpr),
+        TypeError, ("Jaxpr equation LHS .* is ShapedArray(.*), "
+                    "RHS is inferred as ShapedArray(.*), in '.* = sin .*'"))
+
```

Comment:
Can you add a couple tests with (1) subjaxprs (i.e. nested jits), (2) pmaps, and (3) control flow (e.g. `scan`)?

Since we call `check_jaxpr` in several places when we set `core.skip_checks = False`, you could also try running tests with `core.skip_checks = False`, but IIRC they that may fail for unrelated reasons. (It's a TODO to get all tests pass ing with `core.skip_checks = False`...)

Reply:
+1 to enabling checks in tests.

Reply:
George made `core.skip_checks = False` the default for all tests! Wooooo!

Reply:
To clarify the request in my first comment, even though this PR is about checking for errors of first-order primitives only (to start with), we also want to make sure we don't accidentally fail when higher-order primitives are involved (even if we're not catching errors in them).

Since `core.skip_checks = False` now, I suspect that we'll effectively get this kind of when this PR is rebased onto master. So maybe there's no action item here, other than to triple check that this code won't fall down when jit / pmap / control flow is involved.

Reply:
I've rebased. Running all tests with the new default of `skip_checks = False` uncovered a handful of Jaxpr typing bugs, now fixed in this PR. With the fixes in place, all of our tests pass.